### PR TITLE
set up initialization requirements for components

### PR DESCRIPTION
**Pull request by rmudambi**
### Description

- *Category*: feature
- *JIRA issue*: [MIC-4463](https://jira.ihme.washington.edu/browse/MIC-4463)

Enable a Component to properly communicate its dependencies for initialization to the simulation.

### Testing
Wrote a new test that passes.

### CODE REVIEW

File: tests/framework/components/test_component.py [hunk 1]
```diff
@@ -190,6 +204,26 @@ def test_component_that_creates_and_requires_columns_population_view():
...
+        for r in simulation._resource.sorted_nodes
+        # if the producer is an instance method
+        if hasattr(r.producer, "__self__")
+        # and is a method of ColumnCreatorAndRequirer
+        and isinstance(r.producer.__self__, ColumnCreatorAndRequirer)
+    ][0]
+
+    assert "value.pipeline_1" in dependencies
+    assert "column.test_column_2" in dependencies
+    assert "stream.stream_1"
```
Comment (stevebachmeier) [inline tests/framework/components/test_component.py:224]: in dependencies?

File: tests/framework/components/test_component.py [hunk 2]
```diff
@@ -190,6 +204,26 @@ def test_component_that_creates_and_requires_columns_population_view():
...
+    component = ColumnCreatorAndRequirer()
+    simulation = InteractiveContext(components=[ColumnCreator(), component])
+
+    # Assert required resources have been recorded by the ResourceManager
+    dependencies = [
+        r.dependencies
+        # get all resources in the dependency graph
+        for r in simulation._resource.sorted_nodes
+        # if the producer is an instance method
+        if hasattr(r.producer, "__self__")
```
Comment (stevebachmeier) [inline tests/framework/components/test_component.py:217]: Can't you leave this line out and get the same thing with checking it's a method of ColumnCreatorAndRequirer alone?

**Review by hussain-jafari — APPROVED** [review]

File: tests/framework/components/test_component.py (hunk 2 above)
Comment (hussain-jafari) [reply tests/framework/components/test_component.py:217]: Are there resources that aren't instantiated objects?

Comment (rmudambi) [reply tests/framework/components/test_component.py:217]: @hussain-jafari there are, but I don't care about them for this test; I only care about the resource created by `ColumnCreaterAndRequirer`.

**Pull request merged**